selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+  if SHORT:
+    time.sleep(10)
+  else:
+    time.sleep(60 * 10)
+
+def link(src, dest):
+  # Workaround for the EON/termux build of Python having os.link removed.
+  return libc.link(src.encode(), dest.encode())
+
+def run(cmd, cwd=None):
+  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8')
+
+# **** meaty functions ****
+
+def init_ovfs():
+  if os.path.ismount(OVERLAY_MERGED):
+    # We had to have the overlay lock to get this far, so this must be a stale
+    # mount from a terminated updated process.
+    run(["umount", OVERLAY_MERGED])
+
+  cloudlog.info("preparing new staging area")
+  if os.path.isfile(os.path.join(BASEDIR, ".update_succeeded")):
+    os.remove(os.path.join(BASEDIR, ".update_succeeded"))
+  if os.path.isdir(STAGING_ROOT):
+    shutil.rmtree(STAGING_ROOT)
+
+  for dirname in [STAGING_ROOT, OVERLAY_UPPER, OVERLAY_METADATA, OVERLAY_MERGED, FINALIZED]:
+    os.mkdir(dirname, 0o755)
+  if not os.lstat(BASEDIR).st_dev == os.lstat(STAGING_ROOT).st_dev:
+    cloudlog.error("base and staging directories are on different filesystems, not valid for overlay FS!")
+    exit(1)
+
+  overlay_opts = f"lowerdir={BASEDIR},upperdir={OVERLAY_UPPER},workdir={OVERLAY_METADATA}"
+  run(["mount", "-t", "overlay", "-o", overlay_opts, "none", OVERLAY_MERGED])
+
+def inodes_in_tree(search_dir):
+  # Given a search root, produce a dictionary mapping of inodes to relative
+  # pathnames of regular files (no directories, symlinks, or special files).
+  inode_map = {}
+  os.chdir(search_dir)
+  for root, dirs, files in os.walk('.', topdown=True):
+    for file_name in files:
+      full_name = os.path.join(search_dir, root, file_name)
+      st = os.stat(full_name)
+      if S_ISREG(st[ST_MODE]):
+        inode_map.update({st[ST_INO]: full_name})
+  return inode_map
+
+def dup_ovfs_object(inode_map, source_obj, target_dir):
+  # Given a relative pathname to copy, and a new target root, duplicate the
+  # source object in the target root, using hardlinks for regular files.
+
+  st = os.lstat(source_obj)
+  target_obj = os.path.join(target_dir, source_obj)
+
+  if S_ISREG(st[ST_MODE]):
+    # Hardlink all regular files; ownership and permissions are inherited.
+    link(inode_map[st[ST_INO]], target_obj)
+  else:
+    # Recreate all directories and symlinks; copy ownership and permissions.
+    if S_ISDIR(st[ST_MODE]):
+      os.mkdir(os.path.join(FINALIZED, source_obj), S_IMODE(st[ST_MODE]))
+    elif S_ISLNK(st[ST_MODE]):
+      os.symlink(os.readlink(source_obj), target_obj)
+      os.chmod(target_obj, S_IMODE(st[ST_MODE]), follow_symlinks=False)
+    else:
+      # Ran into a FIFO, socket, etc. Should not happen in OP install dir.
+      # Ignore without copying for the time being, revisit if necessary.
+      pass
+    os.chown(target_obj, st[ST_UID], st[ST_GID], follow_symlinks=False)
+    os.utime(target_obj, (st[ST_ATIME], st[ST_MTIME]), follow_symlinks=False)
+
+def finalize_from_ovfs():
```

Comment:
Where is this function called?

Reply:
You caught an interim version. I'm making a minor design change.

I removed it from the main update loop because Git turns out to be overly clever, and somehow notices if the hard-link count increases. So after I finalized, Git would decide everything in OVERLAY_MERGED had been touched and zap it during `git reset --hard`, which broke all the hard-links.

New approach will be to finalize exactly once, at bootup, assuming the success flag is present. I'll do that by having continue.sh call the merged version of updated with a new command-line switch. Finalizing only takes about a second so it won't be a huge contributor to boot time.